contracts/Ethernauts.sol

```diff
@@ -114,6 +121,45 @@ contract Ethernauts is ERC721Enumerable, Ownable {
         return _redeemedCoupons[user];
     }
 
+    function tokenURI(uint256 tokenId) public view override returns (string memory) {
+        string memory baseURI = _baseURI();
+
+        uint batchId = tokenId / batchSize;
+        uint randomNumber = _randomNumberForBatch[batchId];
+        if (randomNumber == 0) {
+            return string(abi.encodePacked(baseURI, "travelling_to_destination"));
+        }
+
+        uint offset = randomNumber % batchSize;
+        uint maxTokenIdInBatch = batchSize * (batchId + 1) - 1;
+
+        uint assetId = tokenId + offset;
+        if (assetId > maxTokenIdInBatch) {
+            assetId -= batchSize;
+        }
+
+        return string(abi.encodePacked(baseURI, assetId.toString()));
+    }
+
+    // This is unprotected for now, but will actually only
+    // be callable by an L1 -> L2 bridge contract.
+    function setRandomNumberForBatch(uint batchId, uint randomNumber) external onlyOwner {
```

Comment:
IMHO, wouldn't be simpler to not receive a `batchId` and calculate it from the previous one? This fn should still check that the next `batchId` is valid, and could have an interface like: 
```solidity
setRandomNumberForNextBatch(uint randomNumber)
```

Reply:
Yeah totally, making the fix...